### DISCUSSION TITLE
"Official" Python 3 Access To A Beets Library

### DISCUSSION THREAD

**Discussion by origintopleft** [beetbox/beets#5073] [Ideas]
I'm writing a plugin for [the IRC bot Limnoria][limnoria] that controls a [LiquidSoap][liq] powered radio stream, effectively letting me override the auto DJ in specific circumstances. The specific function I'm trying to create is a song request command that doesn't require the exact path to a track.

I've chosen Beets as my media library manager for this purpose, since it does exactly what I need it to do (and more I didn't know I needed, thank you!). But in order to make calls to beet, it seems like I would need to use the subprocess module. This would certainly work, but since Limnoria is also Python 3 I would prefer to import beets and directly interface with a "search" submethod.

If I'm not missing something silly here, could I ask for this as a feature request?

[limnoria]: https://github.com/progval/Limnoria
[liq]: https://www.liquidsoap.info

**Comment by Serene-Arc** [comment]
I'm having a little difficulty understanding what you're asking for. Can you elaborate?

The code itself is available but we don't have any API documentation available. You can definitely use the library functions that we have. Also, all of the data is held in a sqlite database, so you can browse that with any tools that work with that kind of data.

**Comment by JOJ0** [comment]
Exactly as @Serene-Arc put it: Beets doesn't have an API like REST you could access (there is something in the queue somewhere but not nearly finished AFAIK). 

Furthermore, if I understand correctly and since I recently also had a use-case like that(simply querying a beets database to get information like path,trackname's,etc) - you don't have to fire subprocess - that is slow, errorprone and actually pretty annoying (been there, done that, in the end  refactored all  to Python)

Accessing a beets library via it's library objects directly is the better faster and actually easier way. I don't have anything of my code in that particular project open-source'd, I'm very sorry :-( but basically start with something like that:

```
from beets.library import Library
from beets.dbcore.query import FieldQuery, AndQuery, OrQuery, SubstringQuery
from beets.ui import print_
from beets.util import syspath, displayable_path, normpath
```

```
+    def _query_lib(beet_query, message, silent=False):
+        """beet_query can be string or query object. Returns a list of paths"""
+        libpath = os.path.expanduser(
+            Path(SHELLENV["BEETSDIR"]) / 'jtbeets_gin.db')
+        lib = Library(libpath)
+        paths_found = []
+
+
+        for item in lib.items(beet_query):
+            paths_found += [item.get('path')]
+        return paths_found
```

Querying examples with that function above:

```
+            beet_result = _query_lib(SubstringQuery('isrc', isrc), '',
+                                     silent=True)
```

```
+                beet_result = _query_lib(
+                    AndQuery([
+                        SubstringQuery('artist', artist_first),
+                        SubstringQuery('title', title)
+                    ]),
```

In my use case I want to query beets library to get fully qualified paths returned. Adapt to your needs! Also, I think there are too much things imported that won't get used in those snippets. I left it, you might need them for something else...

Warning: These are copy paste snippets of my commits and might not work at all. If you are a programme you will get the idea. Please ask back if something is not clear to you! I'll try to help and elaborate. It's been some time since I coded that! ;-)

and then move on with reading beets' developer docs: https://beets.readthedocs.io/en/latest/dev/index.html or maybe even start with the reference: https://beets.readthedocs.io/en/latest/dev/library.html

Does that help somehow?

PS: Sorry for not formatting all that properly. I'm a little busy with things/educational stuff currently. HTH!

**Reply by origintopleft** [reply]
Let me make sure I understand this concept from first principles, because if I do, this is what I meant to ask for.

```
from beets.dbcore.query import SubstringQuery
from beets.library import Library

lib_music = Library("/mnt/samba_share/music.db")
query_response = lib_music.items(SubstringQuery("searchstring"))  # equivalent to: beet ls searchstring
query2_response = lib_music.items(SubstringQuery("artist", "Lacuna Coil")) # equivalent to: beet ls "artist:Lacuna Coil"

# {query,query2}_response is a List-like object containing reasonably-Dict-like objects
```

Am I in the right ballpark? I can't seem to actually find the docs on the `SubstringQuery` object or its relatives (there's a link to Queries in the dev documentation but it points to end user docs again).

**Reply by JOJ0** [reply]
I think that ballpark is the right one :-)) 

Regarding the Query docs, let me have a look...

**Reply by JOJ0** [reply]
OhI see , looks like we have that excluded from Sphinx autodoc 

https://beets.readthedocs.io/en/latest/dev/library.html#queries

As you said backlinking to user docs

Hmm in that case please look it up in the source

https://github.com/beetbox/beets/blob/master/beets/dbcore/query.py

**Reply by add-n2x** [reply]
This is very helpful @JOJ0! Lotta thx.

Do you know how to construct the `beets.library.Library` with an existing Beets configuration file?
